datalad/support/gitrepo.py

```diff
@@ -635,99 +636,178 @@ def _parse_progress_line(self, line):
+                 ' ': FAST_FORWARD,
+                 '=': UP_TO_DATE,
+                 '!': ERROR}
+
+    _operation_map = {
+        NEW_TAG: 'new-tag',
+        NEW_HEAD: 'new-branch',
+        NO_MATCH: 'no-match',
+        REJECTED: 'rejected',
+        REMOTE_REJECTED: 'remote-rejected',
+        REMOTE_FAILURE: 'remote-failure',
+        DELETED: 'deleted',
+        FORCED_UPDATE: 'forced-update',
+        FAST_FORWARD: 'fast-forward',
+        UP_TO_DATE: 'uptodate',
+        ERROR: 'error',
+    }
 
-    def update(self, op_code, cur_count, max_count=None, message=''):
-        # ATM we ignore message which typically includes bandwidth info etc
+    @classmethod
+    def _from_line(cls, line):
+        """Create a new PushInfo instance as parsed from line which is expected to be like
+            refs/heads/master:refs/heads/master 05d2687..1d0568e as bytes"""
+        control_character, from_to, summary = line.split('\t', 3)
+        flags = 0
+
+        # control character handling
         try:
-            if not max_count:
-                # spotted used by GitPython tests, so may be at times it is not
-                # known and assumed to be a 100%...? TODO
-                max_count = 100.0
-            if op_code:
-                # Apparently those are composite and we care only about the ones
-                # we know, so to avoid switching the progress bar for no good
-                # reason - first & with the mask
-                op_code = op_code & self.OP_MASK
-            if self._op_code is None or self._op_code != op_code:
-                # new type of operation
-                self._close_pbar()
-
-                self._pbar = self._ui.get_progressbar(
-                    self._get_human_msg(op_code),
-                    total=max_count,
-                    unit=' objects'
-                )
-                self._op_code = op_code
-            if not self._pbar:
-                lgr.error("Ended up without progress bar... how?")
-                return
-            self._pbar.update(cur_count, increment=False)
-        except Exception as exc:
-            lgr.debug("GitPythonProgressBar errored with %s", exc_str(exc))
-            return
-        #import time; time.sleep(0.001)  # to see that things are actually "moving"
-        # without it we would get only a blink on initial 0 value, istead of
-        # a blink at some higher value.  Anyways git provides those
-        # without flooding so should be safe to force here.
-        self._pbar.refresh()
+            flags |= cls._flag_map[control_character]
+        except KeyError:
+            raise ValueError("Control character %r unknown as parsed from line %r" % (control_character, line))
+        # END handle control character
+
+        # from_to handling
+        from_ref_string, to_ref_string = from_to.split(':')
+
+        # commit handling, could be message or commit info
+        old_commit = None
+        if summary.startswith('['):
```

Comment:
All this string matching of git's porcelain (and potentially translated) output makes me cringe, but this of course is taken from gitpython and is what already happens underneath, so we're not in any worse position.  Plus, I can't think of a way to avoid it, aside from my unpopular opinion that we should relay git's output untouched.

Reply:
I was on the fence whether I should start arguing this way. I decided to provide this implementation this bridge the time between now and a decision that we should make.

If our approach with `publish` and `update` is that we make precise calls that do exactly (and only) what we need -- or -- that we just want to report to a user what was done -- I perfectly agree with you and we should stop with the business ASAP.

If, OTOH, we want to be able to "just push" and then be able to react on it at the datalad-level, we need to keep doing something like this.

This PR is helping to move towards a better `publish()`, and this future one should be able to emit proper results that could be used with hooks. They would require appropriate machine-interpretable reporting.